### PR TITLE
Html rewrite

### PR DESCRIPTION
**Pull request by timstclair**
Update the proxy HTML rewrite logic to avoid building a complete parse tree, instead using the lower level token stream API. Doing so removes the need to create "missing" HTML elements, reducing the collateral changes to the rewritten HTML.

Resolves: #9766
### Benchmarks

The new approach is slightly faster, and scales better with large documents. The documents tested were the kubernetes dashboard index page ("dash", 2.3K), a snapshot of a gmail tab ("mail", 1.7M), and a snapshot of the one-page HTML spec ("html", 9.4M).

**This change**

```
BenchmarkDash      10000        123018 ns/op       26512 B/op        408 allocs/op
BenchmarkMail         50      24182171 ns/op     5638125 B/op      64658 allocs/op
BenchmarkHtml          3     356755878 ns/op    67479413 B/op     901099 allocs/op
```

**Previous method**

```
BenchmarkDash      10000        138123 ns/op       24320 B/op        367 allocs/op
BenchmarkMail         50      25806537 ns/op     5770287 B/op      35871 allocs/op
BenchmarkHtml          2     576100073 ns/op    76078952 B/op     779853 allocs/op
```

### CODE REVIEW

**Comment by timstclair** [comment]
FYI, merged with master so this PR no longer includes the commit from #9756

**Comment by erictune** [comment]
@lavalamp who do you suggest as a reviewer?

File: pkg/util/proxy/transport.go [hunk 1]
```diff
@@ -36,32 +37,32 @@ import (
...
-	"img":        util.NewStringSet("longdesc", "src", "usemap"),
-	"input":      util.NewStringSet("src", "usemap", "formaction"),
-	"ins":        util.NewStringSet("cite"),
-	"link":       util.NewStringSet("href"),
-	"object":     util.NewStringSet("classid", "codebase", "data", "usemap"),
-	"q":          util.NewStringSet("cite"),
-	"script":     util.NewStringSet("src"),
-	"source":     util.NewStringSet("src"),
-	"video":      util.NewStringSet("poster", "src"),
+var tagsToAttrs = map[atom.Atom]util.StringSet{
```
Comment (lavalamp) [inline pkg/util/proxy/transport.go]: atomsToAttrs?

File: pkg/util/proxy/transport.go [hunk 2]
```diff
@@ -139,36 +140,43 @@ func (t *Transport) rewriteURL(targetURL string, sourceURL *url.URL) string {
...
+	var err error
+	for err == nil {
+		tokenType := tokenizer.Next()
+		switch tokenType {
+		case html.ErrorToken:
+			err = tokenizer.Err()
+		case html.StartTagToken, html.SelfClosingTagToken:
+			token := tokenizer.Token()
+			urlAttrs, ok := tagsToAttrs[token.DataAtom]
+			if ok {
```
Comment (lavalamp) [inline pkg/util/proxy/transport.go]: nit: combine w/ previous line: ```if urlAttrs, ok := tagsToAttrs[token.DataAtom]; ok {```

**Comment by lavalamp** [comment]
This is great! Performance improvements are in scope for 1.0. Just two tiny nits.

**Comment by timstclair** [comment]
Thanks Dan. Should I squash the fixes?

**Comment by lavalamp** [comment]
Yes, please! Looks great :)

**Comment by timstclair** [comment]
Done :)

**Comment by lavalamp** [comment]
LGTM

**Comment by saad-ali** [comment]
@k8s-bot ok to test

**Pull request merged**